### PR TITLE
Enhancement/programming exercise/build failed property

### LINKED ISSUE #1153: Make build error identification consistent
You actually can't rely on this string for your check. This is a Bamboo specific string and might be different for other CI systems (With the Jenkins integration, we would get the first problem). I'll open an issue related to this since I saw that we are also using this check in other places.

_Originally posted by @ungaralex in https://github.com/ls1intum/Artemis/pull/1152_

### PR DESCRIPTION
**Pull request by ungaralex**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I implemented the changes with a good performance and prevented too many database calls

### Motivation and Context

Resolves #1153 

We currently rely on a hardcoded String comparison every time we determine whether a result for a submission has a build error (for programmin exercises)

### Description

As discussed with @krusche : 
- I added a new boolean `buildFailed` to the `ProgrammingSubmission` class.
- I migrated the `buildArtifact` boolean to the `ProgrammingSubmission` class, so the `Result` class does not contain this programming exercise specific attribute.

I also added all necessary migration scripts.

### Steps for Testing

1. Log in to Artemis
2. Participate in a programming exercise and produce a build error
3. This status should be reflected in all result components (code-editor, overview, scores, etc.)

### CODE REVIEW

File: src/main/resources/config/liquibase/changelog/20200117213324_changelog.xml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    </changeSet>
+    <changeSet id="20200117213324-2" author="ungar">
+        <sql>
+            update submission
+                set build_artifact = true
+                where discriminator = 'P'
+                and id in (
+                    select submission_id
+                    from result
+                    where build_artifact
```
Comment (krusche) [inline src/main/resources/config/liquibase/changelog/20200117213324_changelog.xml]: = true

File: src/main/java/de/tum/in/www1/artemis/domain/Result.java [hunk 2]
```diff
@@ -59,10 +59,6 @@
     @JsonView(QuizView.After.class)
     private Boolean successful;
 
-    @Column(name = "build_artifact")
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/domain/Result.java]: do not delete, but deprecate it for easier migration

File: src/main/resources/config/liquibase/changelog/20200117213324_changelog.xml [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+                )
+        </sql>
+    </changeSet>
+    <changeSet id="20200117213324-3" author="ungar">
+        <dropColumn tableName="result" columnName="build_artifact" />
```
Comment (krusche) [inline src/main/resources/config/liquibase/changelog/20200117213324_changelog.xml]: remove and execute this changelog some days later for easier migration

File: src/main/java/de/tum/in/www1/artemis/domain/Result.java (hunk 2 above)
Comment (ungaralex) [reply src/main/java/de/tum/in/www1/artemis/domain/Result.java]: I reverted the change

File: src/main/resources/config/liquibase/changelog/20200117213324_changelog.xml (hunk 1 above)
Comment (ungaralex) [reply src/main/resources/config/liquibase/changelog/20200117213324_changelog.xml]: added

File: src/main/resources/config/liquibase/changelog/20200117213324_changelog.xml (hunk 3 above)
Comment (ungaralex) [reply src/main/resources/config/liquibase/changelog/20200117213324_changelog.xml]: Removed the dropColumn change

**Pull request merged**